node/node.go

```diff
@@ -363,66 +367,277 @@ func (node *Node) AddPendingReceipts(receipts *types.CXReceiptsProof) {
 		Msg("Got ONE more receipt message")
 }
 
+type withError struct {
+	err     error
+	payload interface{}
+}
+
+var (
+	errNotRightKeySize = errors.New("key received over wire is wrong size")
+	errNoSenderPubKey  = errors.New("no sender public BLS key in message")
+)
+
+func validateShardBoundMessage(
+	ctx context.Context, payload []byte,
+) (*msg_pb.Message, error) {
+	var (
+		m         msg_pb.Message
+		senderKey *bls.PublicKey
+	)
+
+	if err := protobuf.Unmarshal(payload, &m); err != nil {
+		return nil, errors.WithStack(err)
+	}
+	var senderPubKeyViaWire []byte
+	maybeCon, maybeVC := m.GetConsensus(), m.GetViewchange()
+
+	if maybeCon != nil {
+		senderPubKeyViaWire = maybeCon.GetSenderPubkey()
+	} else if maybeVC != nil {
+		senderPubKeyViaWire = maybeVC.GetSenderPubkey()
+	} else {
+		return nil, errors.WithStack(errNoSenderPubKey)
+	}
+
+	if len(senderPubKeyViaWire) != shard.PublicKeySizeInBytes {
+		return nil, errors.WithStack(errNotRightKeySize)
+	}
+
+	key, err := bls_cosi.BytesToBLSPublicKey(
+		senderPubKeyViaWire,
+	)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	senderKey = key
+
+	if err := consensus.VerifyMessageSig(senderKey, &m); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	return &m, nil
+}
+
 // Start kicks off the node message handling
 func (node *Node) Start() error {
```

Comment:
Should make these method smaller by refactoring out each of the big sections. Or at least add more comment on each sections' purpose. Currently it's very hard to read.

Reply:
will add more functions